federation/pkg/controller/service/servicecontroller.go

```diff
@@ -0,0 +1,949 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package service
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+
+	"reflect"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/api/errors"
+	cache "k8s.io/kubernetes/pkg/client/cache"
+	federationcache "k8s.io/kubernetes/federation/client/cache"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/release_1_3"
+	federationclientset "k8s.io/kubernetes/federation/client/clientset_generated/federation_release_1_3"
+	"k8s.io/kubernetes/pkg/client/record"
+	"k8s.io/kubernetes/pkg/cloudprovider"
+	"k8s.io/kubernetes/pkg/controller"
+	"k8s.io/kubernetes/pkg/controller/framework"
+	pkg_runtime "k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util/runtime"
+	"k8s.io/kubernetes/pkg/util/workqueue"
+	"k8s.io/kubernetes/pkg/util/sets"
+
+	"k8s.io/kubernetes/pkg/watch"
+	"k8s.io/kubernetes/pkg/util/wait"
+	"k8s.io/kubernetes/federation/apis/federation/v1alpha1"
+	"k8s.io/kubernetes/federation/pkg/dnsprovider"
+	//"k8s.io/kubernetes/federation/pkg/dnsprovider/rrstype"
+
+	"k8s.io/kubernetes/pkg/conversion"
+)
+
+const (
+	// To avoid name conflict of service in kubernetes cluster, we would generate random name for the service
+    // before submit the serivce request to kubernetes cluster, and service-spec-hash is the selector name
+	// to link federation service and kubernetes services
+	serviceHashLabelName = "service-spec-hash"
+	serviceSyncPeriod = 30 * time.Second
+	clusterSyncPeriod = 100 * time.Second
+// How long to wait before retrying the processing of a service change.
+// If this changes, the sleep in hack/jenkins/e2e.sh before downing a cluster
+// should be changed appropriately.
+	minRetryDelay = 5 * time.Second
+	maxRetryDelay = 300 * time.Second
```

Comment:
there are two retry mechanism in the servicecontroller, the 1st one is the backoff for whole process retry in case of failure occurring, the 2nd one is when rest call fails, retry 5 times and sleep for 5 second for each attempts(more light weight retry for minor errors.). The logic is copied from k8s service controller and it makes sense to me, do you want them to be merged?